File: 144_NodesDistanceK.cpp
/*
Problem: All Nodes Distance K in Binary Tree
Description: Find all nodes at distance K from target node.

Approach: Build parent pointers, then BFS from target node to find all nodes
          at distance K (treat tree as undirected graph).
*/

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

void buildParentMap(TreeNode* root, unordered_map<TreeNode*, TreeNode*>& parent) {
    if (!root) return;
    
    if (root->left) {
        parent[root->left] = root;
        buildParentMap(root->left, parent);
    }
    
    if (root->right) {
        parent[root->right] = root;
        buildParentMap(root->right, parent);
    }
}

vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
    unordered_map<TreeNode*, TreeNode*> parent;
    buildParentMap(root, parent);
    
    unordered_set<TreeNode*> visited;
    queue<TreeNode*> q;
    q.push(target);
    visited.insert(target);
    
    int distance = 0;
    
    while (!q.empty()) {
        if (distance == k) break;
        
        int size = q.size();
        for (int i = 0; i < size; i++) {
            TreeNode* node = q.front();
            q.pop();
            
            // Check left child
            if (node->left && !visited.count(node->left)) {
                q.push(node->left);
                visited.insert(node->left);
            }
            
            // Check right child
            if (node->right && !visited.count(node->right)) {
                q.push(node->right);
                visited.insert(node->right);
            }
            
            // Check parent
            if (parent.count(node) && !visited.count(parent[node])) {
                q.push(parent[node]);
                visited.insert(parent[node]);
            }
        }
        
        distance++;
    }
    
    vector<int> result;
    while (!q.empty()) {
        result.push_back(q.front()->val);
        q.pop();
    }
    
    return result;
}

int main() {
    TreeNode* root = new TreeNode(3);
    root->left = new TreeNode(5);
    root->right = new TreeNode(1);
    root->left->left = new TreeNode(6);
    root->left->right = new TreeNode(2);
    root->right->left = new TreeNode(0);
    root->right->right = new TreeNode(8);
    root->left->right->left = new TreeNode(7);
    root->left->right->right = new TreeNode(4);
    
    TreeNode* target = root->left;  // Node 5
    int k = 2;
    
    vector<int> result = distanceK(root, target, k);
    
    cout << "Nodes at distance " << k << ": ";
    for (int val : result) {
        cout << val << " ";
    }
    cout << endl;  // Output: 7 4 1
    
    return 0;
}

/*
Time Complexity: O(n)
Space Complexity: O(n)
*/

File: 144_NodesDistanceK.py
"""
Problem: All Nodes Distance K in Binary Tree
Description: Find all nodes at distance K from target node.

Approach: Build parent pointers, then BFS from target node to find all nodes
          at distance K (treat tree as undirected graph).
"""

from typing import Optional, List
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def buildParentMap(root: TreeNode, parent: dict):
    if not root:
        return
    
    if root.left:
        parent[root.left] = root
        buildParentMap(root.left, parent)
    
    if root.right:
        parent[root.right] = root
        buildParentMap(root.right, parent)

def distanceK(root: Optional[TreeNode], target: TreeNode, k: int) -> List[int]:
    parent = {}
    buildParentMap(root, parent)
    
    visited = set()
    q = deque([target])
    visited.add(target)
    
    distance = 0
    
    while q:
        if distance == k:
            break
        
        size = len(q)
        for _ in range(size):
            node = q.popleft()
            
            # Check left child
            if node.left and node.left not in visited:
                q.append(node.left)
                visited.add(node.left)
            
            # Check right child
            if node.right and node.right not in visited:
                q.append(node.right)
                visited.add(node.right)
            
            # Check parent
            if node in parent and parent[node] not in visited:
                q.append(parent[node])
                visited.add(parent[node])
        
        distance += 1
    
    return [node.val for node in q]

def main():
    root = TreeNode(3)
    root.left = TreeNode(5)
    root.right = TreeNode(1)
    root.left.left = TreeNode(6)
    root.left.right = TreeNode(2)
    root.right.left = TreeNode(0)
    root.right.right = TreeNode(8)
    root.left.right.left = TreeNode(7)
    root.left.right.right = TreeNode(4)
    
    target = root.left  # Node 5
    k = 2
    
    result = distanceK(root, target, k)
    
    print(f"Nodes at distance {k}:", result)  # Output: [7, 4, 1]

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
